### ISSUE TITLE
Tokenization mismatch while finetuning LLaVA-v1.6-34B

### ISSUE THREAD

**Issue by Bleking** [haotian-liu/LLaVA#1706]
### Describe the issue

I tried to LoRA finetune LLaVA-v1.6-34B with my own dataset(size 1000) and I constantly face tokenization mismatch during the process.

This is my script for LLaVA-v1.6-34B, `finetune_task_lora.sh`. I will not show the data_path and image_folder for my privacy.
```
#!/bin/bash

deepspeed llava/train/train_xformers.py \
    --lora_enable True --lora_r 128 --lora_alpha 256 --mm_projector_lr 2e-5 \
    --deepspeed ./scripts/zero3_offload.json \
    --model_name_or_path ./checkpoints/pretrained/llava-v1.6-34b \
    --version v1 \
    --data_path ./playground/data/- \
    --image_folder ./playground/data/- \
    --vision_tower openai/clip-vit-large-patch14-336 \
    --mm_projector_type mlp2x_gelu \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end False \
    --mm_use_im_patch_token False \
    --image_aspect_ratio pad \
    --group_by_modality_length True \
    --fp16 True \
    --output_dir ./checkpoints/llava-v1.6-vicuna-34b-task-lora-floorplan-vqa-1000 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 4 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 8 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 50000 \
    --save_total_limit 1 \
    --learning_rate 2e-4 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 False \
    --model_max_length 4096 \
    --gradient_checkpointing True \
    --dataloader_num_workers 0 \
    --lazy_preprocess True \
    --report_to wandb
```

Also, checked the tokenizer defined by `AutoTokenizer.from_pretrained('/home/work/testdataset1/LLaVA/checkpoints/pretrained/llava-v1.6-34b')` and I got the following:
```
LlamaTokenizerFast(name_or_path='/home/work/testdataset1/LLaVA/checkpoints/pretrained/llava-v1.6-34b', vocab_size=64000, model_max_length=4096, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<|startoftext|>', 'eos_token': '<|im_end|>', 'unk_token': '<unk>', 'pad_token': '<unk>'}, clean_up_tokenization_spaces=False),  added_tokens_decoder={
        0: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        1: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        6: AddedToken("<|im_start|>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=False),
        7: AddedToken("<|im_end|>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        64000: AddedToken("<|startoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        64001: AddedToken("<|endoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
}
```
and set model_max_length=4096 following this result

I will also share with you the list of modules in the 'llava' environment I use for this.
```
Package                       Version
----------------------------- -----------
accelerate                    0.21.0
aiofiles                      23.2.1
aiohappyeyeballs              2.4.0
aiohttp                       3.10.5
aiosignal                     1.3.1
altair                        5.4.1
annotated-types               0.7.0
anyio                         4.4.0
async-timeout                 4.0.3
attrs                         24.2.0
av                            13.0.0
bitsandbytes                  0.43.3
black                         24.1.0
certifi                       2024.8.30
cfgv                          3.4.0
chardet                       5.2.0
charset-normalizer            3.3.2
click                         8.1.7
contourpy                     1.3.0
cycler                        0.12.1
DataProperty                  1.0.1
datasets                      2.16.1
decord                        0.6.0
deepspeed                     0.14.4
dill                          0.3.7
distlib                       0.3.8
distro                        1.9.0
docker-pycreds                0.4.0
einops                        0.6.1
einops-exts                   0.0.4
et-xmlfile                    1.1.0
evaluate                      0.4.3
exceptiongroup                1.2.2
fastapi                       0.115.0
ffmpy                         0.4.0
filelock                      3.16.1
flash-attn                    2.6.3
fonttools                     4.53.1
frozenlist                    1.4.1
fsspec                        2023.10.0
ftfy                          6.2.3
gitdb                         4.0.11
GitPython                     3.1.43
gradio                        4.16.0
gradio_client                 0.8.1
h11                           0.14.0
hf_transfer                   0.1.8
hjson                         3.1.0
httpcore                      0.17.3
httpx                         0.24.0
huggingface-hub               0.25.0
identify                      2.6.1
idna                          3.10
importlib_resources           6.4.5
isort                         5.13.2
Jinja2                        3.1.4
jiter                         0.5.0
joblib                        1.4.2
jsonlines                     4.0.0
jsonschema                    4.23.0
jsonschema-specifications     2023.12.1
kiwisolver                    1.4.7
latex2mathml                  3.77.0
llava                         1.2.2.post1
lmms_eval                     0.2.3.post1
loguru                        0.7.2
markdown-it-py                3.0.0
markdown2                     2.5.0
MarkupSafe                    2.1.5
matplotlib                    3.9.2
mbstrdecoder                  1.1.3
mdurl                         0.1.2
mpmath                        1.3.0
multidict                     6.1.0
multiprocess                  0.70.15
mutagen                       1.47.0
narwhals                      1.8.1
networkx                      3.3
ninja                         1.11.1.1
nodeenv                       1.9.1
numpy                         1.26.4
nvidia-cublas-cu12            12.1.3.1
nvidia-cuda-cupti-cu12        12.1.105
nvidia-cuda-nvrtc-cu12        12.1.105
nvidia-cuda-runtime-cu12      12.1.105
nvidia-cudnn-cu12             8.9.2.26
nvidia-cufft-cu12             11.0.2.54
nvidia-curand-cu12            10.3.2.106
nvidia-cusolver-cu12          11.4.5.107
nvidia-cusparse-cu12          12.1.0.106
nvidia-ml-py                  12.560.30
nvidia-nccl-cu12              2.18.1
nvidia-nvjitlink-cu12         12.6.68
nvidia-nvtx-cu12              12.1.105
openai                        1.46.0
openpyxl                      3.1.5
orjson                        3.10.7
packaging                     24.1
pandas                        2.2.2
pathvalidate                  3.2.1
peft                          0.12.0
pillow                        10.4.0
pip                           24.2
platformdirs                  4.3.6
portalocker                   2.10.1
pre-commit                    3.8.0
protobuf                      3.20.0
psutil                        6.0.0
py-cpuinfo                    9.0.0
pyarrow                       17.0.0
pyarrow-hotfix                0.6
pycocoevalcap                 1.2
pycocotools                   2.0.8
pycryptodomex                 3.20.0
pydantic                      2.9.2
pydantic_core                 2.23.4
pydub                         0.25.1
Pygments                      2.18.0
pynvml                        11.5.3
pyparsing                     3.1.4
pytablewriter                 1.2.0
python-dateutil               2.9.0.post0
python-multipart              0.0.9
pytz                          2024.2
PyYAML                        6.0.2
referencing                   0.35.1
regex                         2024.9.11
requests                      2.32.3
rfc3986                       1.5.0
rich                          13.8.1
rpds-py                       0.20.0
ruff                          0.6.5
sacrebleu                     2.4.3
safetensors                   0.4.5
scikit-learn                  1.2.2
scipy                         1.14.1
semantic-version              2.10.0
sentencepiece                 0.1.99
sentry-sdk                    2.14.0
setproctitle                  1.3.3
setuptools                    72.1.0
shellingham                   1.5.4
shortuuid                     1.0.13
six                           1.16.0
smmap                         5.0.1
sniffio                       1.3.1
sqlitedict                    2.1.0
starlette                     0.38.5
svgwrite                      1.4.3
sympy                         1.13.2
tabledata                     1.3.3
tcolorpy                      0.1.6
tenacity                      8.3.0
threadpoolctl                 3.5.0
tiktoken                      0.7.0
timm                          0.6.13
tokenizers                    0.15.1
tomlkit                       0.12.0
torch                         2.1.2
torchaudio                    2.1.2+cu121
torchvision                   0.16.2
tqdm                          4.66.5
tqdm-multiprocess             0.0.11
transformers                  4.37.2
transformers-stream-generator 0.0.5
triton                        2.1.0
typepy                        1.3.2
typer                         0.12.5
typing_extensions             4.12.2
tzdata                        2024.1
urllib3                       2.2.3
uvicorn                       0.30.6
virtualenv                    20.26.5
wandb                         0.18.1
wavedrom                      2.0.3.post3
websockets                    11.0.3
wheel                         0.44.0
xxhash                        3.5.0
yarl                          1.11.1
yt-dlp                        2024.8.6
zss                           1.2.0
zstandard                     0.23.0
```

Regarding my script and the module versions, what would be the causes of the tokenization mismatches?
Because of this, the loss stay 0.0 during the whole finetuning process.

I will share only a small part of the mismatch messages in case you need some.
WARNING: tokenization mismatch: 385 vs. 399. (ignored)
WARNING: tokenization mismatch: 381 vs. 395. (ignored)
WARNING: tokenization mismatch: 207 vs. 221. (ignored)
WARNING: tokenization mismatch: 234 vs. 248. (ignored)
WARNING: tokenization mismatch: 187 vs. 201. (ignored)
WARNING: tokenization mismatch: 188 vs. 202. (ignored)
WARNING: tokenization mismatch: 159 vs. 173. (ignored)
WARNING: tokenization mismatch: 371 vs. 385. (ignored)
WARNING: tokenization mismatch: 149 vs. 163. (ignored)
WARNING: tokenization mismatch: 210 vs. 224. (ignored)
WARNING: tokenization mismatch: 357 vs. 371. (ignored)
WARNING: tokenization mismatch: 195 vs. 209. (ignored)
WARNING: tokenization mismatch: 169 vs. 183. (ignored)
WARNING: tokenization mismatch: 234 vs. 248. (ignored)
WARNING: tokenization mismatch: 386 vs. 400. (ignored)
WARNING: tokenization mismatch: 204 vs. 218. (ignored)

Thank you.

**Comment by Bleking** [comment]
I solved the issue by changing the 'version' value in the [script](https://github.com/haotian-liu/LLaVA/blob/main/scripts/v1_5/finetune_task_lora.sh) as 'mpt' and initialising `cur_len = 1` in the "preprocess_mpt" function, because I am using 'mpt' instead of 'v1'. (I used this script as I have been finetuning the model with my custom dataset.)

It is because, according to this [tokenizer_config.json](https://huggingface.co/liuhaotian/llava-v1.6-34b/blob/main/tokenizer_config.json), the eos_token of v1.6-34B is `<|im_end|>`, not `</s>`, according to this [tokenizer_config.json](https://huggingface.co/liuhaotian/llava-v1.6-34b/blob/main/tokenizer_config.json). We have to check the way of tokenization carefully edit the script accordingly.

For anyone facing the similar issue, I hope this was helpful! ;)
